x-vector/src/X/Data/Vector/Stream/Group.hs

```diff
@@ -32,39 +33,60 @@ concatWith =
   flip Stream.concatMap
 {-# INLINE [1] concatWith #-}
 
-chunksOf :: (PrimMonad b, MonadBase b m, Generic.Vector v a) => Int -> Stream m a -> Stream m (v a)
-chunksOf !n (Stream step t) =
+-- | Turns a stream of @a@ in to a stream of chunks of size @n@.
+chunked :: (PrimMonad b, MonadBase b m, Generic.Vector v a) => Int -> Stream m a -> Stream m (v a)
+chunked n xs =
   if n <= 0 then
-    Savage.error "X.Data.Vector.Stream.Group.chunksOf: chunk size must be greater than zero"
+    Savage.error "X.Data.Vector.Stream.Group.chunked: chunk size must be greater than zero"
+  else
+    grouped (\_ _ -> False) n xs
+{-# INLINE [1] chunked #-}
+
+-- | Turns a stream of @a@ in to a stream of chunks of at least size @n@,
+--   except for the last one. Values of @a@ which are equal according to the
+--   comparison function stay in the same chunk.
+grouped :: (PrimMonad b, MonadBase b m, Generic.Vector v a) => (a -> a -> Bool) -> Int -> Stream m a -> Stream m (v a)
+grouped eq n (Stream step t) =
+  if n <= 0 then
+    Savage.error "X.Data.Vector.Stream.Group.grouped: chunk size must be greater than zero"
   else
     embed $ do
```

Comment:
I've got a bad feeling about `embed`. if this ever ends up being a bottleneck, I suspect adding a new "uninitialised" state to defer constructing the first grow vector.
might not actually be an issue though